source/common/common/fmt.h

```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include "envoy/common/platform.h" // Avert format.h including windows.h
```

Comment:
Can we fix `fmtlib` upstream?

Reply:
As a practical matter? Any of our 3 dozen dependencies could introduce an #include <window.h> and disturb our source compilation, that header is not meant for general C++ language consumption. By this, we mean that MS heavily leverages #define and C++ heavily leverages multiple legible tokens. The logic in C language might introduce add_interface() but in C++ we are going to introduce interface.add(), which collides with windows.h defines.

We've introduced platform.h to work around this every time we encounter the issue. And if it is fixed upstream, then we could drop our #include, but it isn't critical. The answer at envoy is to eventually stop using Posix or Windows types in our own .h interface definitions, and localize the Posix and Windows implementations to our .cc sources (which will always include platform.h.) Exceptions would include our .h interface stubs to third party libraries such as grpc, nghttp2, etc.

I don't believe an upstream fix is really a priority, since I'm reading on #envoy-dev that fmtlib is being phased out for absl? It certainly shouldn't block this patch.

Reply:
I feel that any 1st class support for Windows in Envoy doesn't end at Envoy. With my maintainer hat on I'm labeling this technical debt.. It's just a single line, but one day someone will come along and spend a day or so chasing down some obscure issue that is masked by this.

At the same time, we're pragmatic about tech debt and want to see Windows support land.

I would like to see tracking issues filed with the respective communities or in Envoy GH and referenced in these comments. Even if you folks aren't going to own this, we need to recognize these problems and provide the opportunity for other folks to do so.

Reply:
Also, we will live with `fmtlib` for a long time. It's deprecated on the data path, but it's a key part of logging which won't be going away.

Reply:
> Also, we will live with fmtlib for a long time. It's deprecated on the data path, but it's a key part of logging which won't be going away.

Fair point, and researching, it turns out they have already responded.

https://github.com/fmtlib/fmt/issues/171

We have the option of disabling fmt.h's  use of ``windows.h``, we just need to define ``FMT_USE_WINDOWS_H=0`` as appropriate. This simply deprives us of the optimizations present in bazel-envoy/external/com_github_fmtlib_fmt/include/fmt/format-inl.h

It's an open question whether we want to define this symbol prior to including format.h in our 
source/common/common/fmt.h header or include windows.h before they can, these have the same effect. If we want to explore this further, perhaps an issue is in order for further discussions, but we are reasonably happy with the proposed solution.